src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -41,21 +44,61 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
+            // While we do not expect to run in to this scenario very often, we'd still want to minimize the impact on user experience of this feature
+            // as best as we could when it does occur. So the solution we came up with is this: if we decided to not include expanded items (because the
+            // computation is running too long,) we will let it run in the background as long as the completion session is still active. Then whenever
+            // any user input that would cause the completion list to refresh, we will check the state of this background task and add expanded items as part
+            // of the update if they are available.
+            // There is a `CompletionContext.IsIncomplete` flag, which is only supported in LSP mode at the moment. Therefore we opt to handle the checking
+            // and combining the items in Roslyn until the `IsIncomplete` flag is fully supported in classic mode.
+
+            if (session.Properties.TryGetProperty<Task<(CompletionContext, RoslynCompletionList)>>(CompletionSource.ExpandedItemsTask, out var task)
+                && task.Status == TaskStatus.RanToCompletion)
+            {
+                // Make sure the task is removed when Adding expanded items,
+                // so duplicated items won't be added in subsequent list updates.
+                session.Properties.RemoveProperty(CompletionSource.ExpandedItemsTask);
+
+                var (expandedContext, _) = await task.ConfigureAwait(false);
+                if (expandedContext.Items.Length > 0)
+                {
+                    // Here we rely on the implementation detail of `CompletionItem.CompareTo`, which always put expand items after regular ones.
+                    var _ = ArrayBuilder<VSCompletionItem>.GetInstance(expandedContext.Items.Length + data.InitialSortedList.Length, out var itemsBuilder);
+                    itemsBuilder.AddRange(data.InitialSortedList);
+                    itemsBuilder.AddRange(expandedContext.Items);
+                    var combinedList = itemsBuilder.ToImmutable();
+
+                    // Add expanded items into a combined list, and save it to be used for future updates during the same session.
+                    session.Properties[CombinedSortedList] = combinedList;
+                    var combinedFilterStates = FilterSet.CombineFilterStates(expandedContext.Filters, data.SelectedFilters);
+
+                    data = new(combinedList, data.Snapshot, data.Trigger, data.InitialTrigger, combinedFilterStates,
+                        data.IsSoftSelected, data.DisplaySuggestionItem, data.Defaults);
+                }
+
+                AsyncCompletionLogger.LogSessionWithDelayedImportCompletionIncludedInUpdate();
+            }
+            else if (session.Properties.TryGetProperty<ImmutableArray<VSCompletionItem>>(CombinedSortedList, out var combinedSortedList))
+            {
+                // Always use the previously saved combined list if available.
```

Comment:
i don't understand ths... why are we even putting this into the properties... instead of just having this be in our model?